src/qibo/tomography/gate_set_tomography.py

```diff
@@ -0,0 +1,277 @@
+from functools import cache
+from inspect import signature
+from itertools import product
+from typing import Union
+
+import numpy as np
+from sympy import S
+
+from qibo import Circuit, gates, symbols
+from qibo.backends import GlobalBackend
+from qibo.config import raise_error
+from qibo.hamiltonians import SymbolicHamiltonian
+
+
+@cache
+def _gates(nqubits):
+    return list(
+        product(
+            [(gates.I,), (gates.X,), (gates.H,), (gates.H, gates.S)], repeat=nqubits
+        )
+    )
+
+
+@cache
+def _measurements(nqubits):
+    return list(product([gates.Z, gates.X, gates.Y, gates.Z], repeat=nqubits))
+
+
+@cache
+def _observables(nqubits):
+    return list(product([symbols.I, symbols.Z, symbols.Z, symbols.Z], repeat=nqubits))
+
+
+@cache
+def _get_observable(j, nqubits):
```

Comment:
I notice that there are functions which I didn't write like `_gates(nqubits)`, `_measurements(nqubits)`, `_observables(nqubits)`, and `_get_observable(j, nqubits)`. Can I suggest that the person who adds these functions add the docstring? It's much easier that way so that we waste less time figuring out what the functions do. Since these are functions listed above are just a few lines, I will add them.

Reply:
I created those functions for better readability and improving the performance by caching them, but I did not change much implementation wise, i.e. they are doing what they originally were, just separately.